### PR TITLE
Add function and tests for encoder function

### PR DESCRIPTION
**Pull request by michaelavs**
@erogluorhan has taken this on as @michaelavs has left NCAR.

Add `encode_as` function and update relevant scrip, exodus, and ugrid encode functions as well the grid module, test scripts, and documentation files accordingly.

In light of the recent Grid class design changes we are planning, this PR changes the `write` function to an `encode_as` function in the UXarray PR and leaves the task of writing out to files (netcdf or zarr) to xarray. cc: @rajeeja @rljacob 

Changes:
-`encode_as` function in `grid.py` returns a `xarray.Dataset` in the given grid type and no longer writes that dataset to a new output file (Leaves this task tot the user's calls such as `uxarray.Grid.encode_as(...).to_netcdf()` or `...to_zarr()`)
- Update documentation including the draft UXarray API accordingly

### CODE REVIEW

**Comment by michaelavs** [comment]
@erogluorhan @anissa111 
For the `to_zarr` issue, I get the issue when changing the tests in `test_grid` from "netcdf" to "zarr". The the error is `NameError: name 'zarr' is not defined`. I'm not sure if this is from using the grid object or not, as I believe that should still be in a xaray.Dataset and not uxarray.grid. The documentation for the xarray function can be found here: https://docs.xarray.dev/en/stable/generated/xarray.Dataset.to_zarr.html

File: uxarray/grid.py [hunk 1]
```diff
@@ -160,33 +160,69 @@ def __from_ds__(self, dataset):
...
+        Parameters
+        ----------
+        outfile : str, required
+
+        file_type : str, required
+            New file type to be created from write function. Supported options are
+            "netcdf" and "zarr"
+        """
+        if file_type == "netcdf":
+            self.ds.to_netcdf(outfile)
```
Comment (erogluorhan) [inline uxarray/grid.py:176]: Intention is not to write the `self.ds` to the output files. Our aim is to write out the dataset that was returned out of the encoders that got in `self.ds`.

**Review by erogluorhan — CHANGES_REQUESTED** [review]
Thanks for this! Please address the inline comments!

**Comment by erogluorhan** [comment]
> @erogluorhan @anissa111 For the `to_zarr` issue, I get the issue when changing the tests in `test_grid` from "netcdf" to "zarr". The the error is `NameError: name 'zarr' is not defined`. I'm not sure if this is from using the grid object or not, as I believe that should still be in a xaray.Dataset and not uxarray.grid. The documentation for the xarray function can be found here: https://docs.xarray.dev/en/stable/generated/xarray.Dataset.to_zarr.html

I'll look into it.

File: uxarray/grid.py [hunk 2]
```diff
@@ -160,33 +160,57 @@ def __from_ds__(self, dataset):
...
 
+        save_as : str, default "netcdf"
+            The specific file type to save newly created datasets to.
+            Current options are "netcdf", "zarr", and "None" (no file saved)
+
         Raises
         ------
         RuntimeError
-            If unsupported grid type provided or directory not found
+            If unsupported provided grid type, file type, or outfile directory is not found
```
Comment (erogluorhan) [inline uxarray/grid.py:186]: The newer phrase seems to be less readable

File: uxarray/grid.py [hunk 3]
```diff
@@ -160,33 +160,57 @@ def __from_ds__(self, dataset):
...
         elif grid_type == "scrip":
-            _write_scrip(outfile, self.Mesh2_face_nodes, self.Mesh2_node_x,
-                         self.Mesh2_node_y, self.face_areas)
+            out_ds = _encode_scrip(self.Mesh2_face_nodes, self.Mesh2_node_x,
+                                   self.Mesh2_node_y, self.face_areas)
         else:
             raise RuntimeError("Format not supported for writing: ", grid_type)
 
+        if save_as == 'netcdf':
+            out_ds.to_netcdf(outfile)
```
Comment (erogluorhan) [inline uxarray/grid.py:202]: I think we need to find a way to support the whole signature of Xarray.Dataset.to_netcdf (`Dataset.to_netcdf(path=None, mode='w', format=None, group=None, engine=None, encoding=None, unlimited_dims=None, compute=True, invalid_netcdf=False)`)

File: uxarray/grid.py [hunk 4]
```diff
@@ -160,33 +160,57 @@ def __from_ds__(self, dataset):
             raise RuntimeError("unknown file format: " + self.mesh_filetype)
         dataset.close()
 
-    def write(self, outfile, grid_type):
+    def write(self, outfile, grid_type, save_as='netcdf'):
         """Writes mesh file as per extension supplied in the outfile string.
```
Comment (erogluorhan) [inline uxarray/grid.py:164]: As per extension in the outfile string or the save_as format?

File: uxarray/grid.py [hunk 5]
```diff
@@ -160,33 +160,57 @@ def __from_ds__(self, dataset):
...
+            ex: "path/to/new/file/location/new_file.nc" or
+                "new_file.nc"
+
         grid_type : str, required
             Grid type of output file.
             Currently supported options are "ugrid", "exodus", and "scrip"
 
+        save_as : str, default "netcdf"
+            The specific file type to save newly created datasets to.
+            Current options are "netcdf", "zarr", and "None" (no file saved)
```
Comment (erogluorhan) [inline uxarray/grid.py:181]: As this is a write function, is there any use case with no file save (i.e. save_as=="None")?

Comment (erogluorhan) [inline uxarray/grid.py:181]: `zarr` package needs to be added into dependencies (also shouldn't be forgotten to be added in the conda-forge recipe). It is the reason of `to_zarr` call not working correctly.

File: uxarray/grid.py [hunk 6]
```diff
@@ -160,33 +160,57 @@ def __from_ds__(self, dataset):
...
+            out_ds = _encode_scrip(self.Mesh2_face_nodes, self.Mesh2_node_x,
+                                   self.Mesh2_node_y, self.face_areas)
         else:
             raise RuntimeError("Format not supported for writing: ", grid_type)
 
+        if save_as == 'netcdf':
+            out_ds.to_netcdf(outfile)
+
+        elif save_as == 'zarr':
+            out_ds.to_zarr()
```
Comment (erogluorhan) [inline uxarray/grid.py:205]: Same with the previous comment. Should support everything `xarray.Dataset.to_zarr` supports.

File: uxarray/_scrip.py [hunk 7]
```diff
@@ -138,8 +138,7 @@ def _read_scrip(ext_ds):
     return ds
 
 
-def _write_scrip(outfile, mesh2_face_nodes, mesh2_node_x, mesh2_node_y,
-                 face_areas):
+def _encode_scrip(mesh2_face_nodes, mesh2_node_x, mesh2_node_y, face_areas):
     """Function to reassign UGRID formatted variables to SCRIP formatted
     variables and then writing them out to a netCDF file.
```
Comment (erogluorhan) [inline uxarray/_scrip.py:143]: "Writing to a netCDF file" is no longer this function's task.

File: uxarray/_exodus.py [hunk 8]
```diff
@@ -141,19 +141,20 @@ def _read_exodus(ext_ds, ds_var_names):
     return ds
 
 
-def _write_exodus(ds, outfile, ds_var_names):
+def _encode_exodus(ds, outfile, ds_var_names):
     """Exodus file writer.
```
Comment (erogluorhan) [inline uxarray/_exodus.py:145]: "File writing" is no longer this function's task.

File: uxarray/_ugrid.py [hunk 9]
```diff
@@ -54,17 +54,13 @@ def _read_ugrid(xr_ds, var_names_dict):
 
 
 # Write a uxgrid to a file with specified format.
-def _write_ugrid(ds, outfile, ugrid_vars):
+def _encode_ugrid(ds):
     """UGRID file writer.
```
Comment (erogluorhan) [inline uxarray/_ugrid.py:58]: "File writing" is no longer this function's task.

File: uxarray/_exodus.py [hunk 10]
```diff
@@ -141,19 +141,20 @@ def _read_exodus(ext_ds, ds_var_names):
     return ds
 
 
-def _write_exodus(ds, outfile, ds_var_names):
+def _encode_exodus(ds, outfile, ds_var_names):
```
Comment (erogluorhan) [inline uxarray/_exodus.py:144]: `outfile` does not need to be an argument anymore

File: test/test_grid.py [hunk 11]
```diff
@@ -49,6 +49,23 @@ def test_read_ugrid_write_scrip(self):
...
+
+        # Create paths for new scrip and exodus files to be stored
+        scrip_outfile = current_path / "meshfiles" / "test_scrip_to_netcdf.nc"
+        exodus_outfile = current_path / "meshfiles" / "test_exodus_to_netcdf.nc"
+        ugrid_outfile = current_path / "meshfiles" / "test_ugrid_to_netcdf.nc"
+
+        # User writer function with encoder argument
+        tgrid1.write(str(scrip_outfile), "scrip", 'netcdf')
+        tgrid1.write(str(exodus_outfile), "exodus", "netcdf")
+        tgrid1.write(str(ugrid_outfile), "ugrid", "netcdf")
```
Comment (erogluorhan) [inline test/test_grid.py:67]: `zarr` case needs to be tested, too.

**Review by erogluorhan — COMMENTED** [review]
Thanks for improvements! Please address inline comments.

File: uxarray/_exodus.py (hunk 10 above)
Comment (michaelavs) [reply uxarray/_exodus.py:144]: `outfile` is used in this function as a title for the dataset. Removing it doesn't change the overall funcitonality, however, I wasn't sure if removing it (and corresponding uses for the title) would be what was/is wanted and if the title in the dataset was required for following exodus standards or was just something we added

File: uxarray/grid.py (hunk 3 above)
Comment (dcherian) [reply uxarray/grid.py:202]: IMO the user could easily do `grid.encode_as("scrip").to_netcdf(...)` where `grid.encode` returns a Dataset. That means uxarray doesn't have to deal with all the details and file format backends that keep changing

Comment (erogluorhan) [reply uxarray/grid.py]: I agree. @michaelavs please let me know if you can do this tomorrow; otherwise, I can do it.

Comment (erogluorhan) [reply uxarray/grid.py]: I have made it this way

**Comment by erogluorhan** [comment]
I have taken this on as @michaelavs has left NCAR.

In light of the recent Grid class design changes we are planning, this PR changes the `write` function to an `encode_as` function in the UXarray PR and leaves the task of writing out to files (netcdf or zarr) to xarray. Please see the PR description for details.

**Review by erogluorhan — APPROVED** [review]

**Review by rljacob — APPROVED** [review]

**Pull request merged**